modules/replication/module.py

```diff
@@ -55,18 +61,162 @@ def __init__(self, id, resolver, n, f, k):
+
+            # line 15 - 25
+            if (self.resolver.execute(
+                    Module.VIEW_ESTABLISHMENT_MODULE,
+                    Function.ALLOW_SERVICE) and (self.need_flush is False)):
+                if self.resolver.execute(
+                    Module.PRIMARY_MONITORING_MODULE,
+                    Function.NO_VIEW_CHANGE) \
+                        and self.rep[self.id][VIEW_CHANGE] is False:
+                    if self.id == self.prim:
+                        # primary processes all pending reqs
+                        for req in self.get_pend_reqs():
+                            if self.seq_n < (self.last_exec() +
+                               (SIGMA * self.number_of_clients)):
+                                self.seq_n += 1
+                                req = {
+                                    CLIENT_REQ: req,
+                                    VIEW: self.prim,
+                                    SEQUENCE_NO: self.seq_n
+                                }
+                                self.rep[self.id][REQ_Q].append({
+                                    REQUEST: req,
+                                    STATUS: ReplicationEnums.PRE_PREP
+                                })
+                                self.rep[self.id][REQ_Q].append({
+                                    REQUEST: req,
+                                    STATUS: ReplicationEnums.PREP
+                                })
+
+                    else:
+                        # wait for prim or process reqs where 3f+1
+                        # agree on seqnum
+                        reqs = list(filter(
+                            self.reqs_to_prep, self.known_pend_reqs()))
+                        for r in reqs:
+                            self.rep[self.id][REQ_Q].append(
+                                {REQUEST: r, STATUS: ReplicationEnums.PREP}
+                            )
+
+                    # consider prepped msgs per request,
+                    # if 3f+1 agree then commit
+                    for req in self.known_reqs(ReplicationEnums.PREP):
+                        req[STATUS] = ReplicationEnums.COMMIT
+                        self.rep[self.id][PEND_REQS].remove(req)
+
+                    for req in self.known_reqs(
+                            set([ReplicationEnums.PREP,
+                                 ReplicationEnums.COMMIT])):
+                        x_set = self.committed_set(req)
+                        if (len(x_set) > (3 * self.number_of_byzantine) + 1) \
+                                and (req[SEQUENCE_NO] == self.last_exec() + 1):
+                            self.commit({REQUEST: req, X_SET: x_set})
+
+    def reqs_to_prep(self, req):
+        """Helper method to filter out requests to prepare."""
+        if req in self.unassigned_reqs():
+            return False
+        for replica_structure in self.rep:
+            if req in replica_structure[REQ_Q]:
```

Comment:
known_pend_reqs() will give back a request, which will be `req`, in rep[REQ_Q] the structure is {REQUEST: req, STATUS: status}. which means that we need to check:
```
for rep_pair in replica_structure[REQ_Q]:
    if req == rep_pair[REQUEST]:
        return False
```